FF1Lib/PredictivePlacement.cs

```diff
@@ -0,0 +1,447 @@
+			var preBlackOrbLocationPool = _incentivesData.AllValidPreBlackOrbItemLocations.ToList();
+			var preBlackOrbUnincentivizedLocationPool = preBlackOrbLocationPool.Where(x => !incentiveLocationPool.Any(y => y.Address == x.Address)).ToList();
+			if ((bool)_flags.LooseExcludePlacedDungeons)
+				preBlackOrbUnincentivizedLocationPool = IncentivizedDungeons(preBlackOrbUnincentivizedLocationPool);
+
+			var unincentivizedLocationPool = new HashSet<IRewardSource>(preBlackOrbUnincentivizedLocationPool, new RewardSourceEqualityComparer());
+
+			List<IRewardSource> placedItems = null;
+			List<Item> treasurePool = null;
+
+			bool placementFailed;
+			do
+			{
+				placementFailed = false;
+
+				var balancedPicker = new RewardSourcePicker(0.5, _flags.LooseItemsNpcBalance ? 7.0 : 1.0, _checker);
+
+				_sanityCounter++;
+				if (_sanityCounter > 2) throw new InsaneException("Item Placement could not meet incentivization requirements!");
+
+				placedItems = ctx.Forced.ToList();
+				var incentives = new HashSet<Item>(ctx.Incentivized);
+				var nonincentives = ctx.Unincentivized.ToList();
+				var shards = ctx.Shards.ToList();
+				treasurePool = ctx.AllTreasures.ToList();
+				var state = PlacementState.Normal;
+
+				while (incentives.Count() > incentiveLocationPool.Count())
+				{
+					nonincentives.Add(incentives.SpliceRandom(rng));
+				}
+
+				if (((bool)_flags.NPCItems) || ((bool)_flags.NPCFetchItems))
+				{
+					HashSet<Item> allPlacements = new HashSet<Item>(nonincentives.Concat(incentives));
+					HashSet<Item> allKeyItems = new HashSet<Item>(MapChangeItems.Concat(FetchQuestItems).Concat(GatingItems));
+
+					if ((bool)_flags.IsFloaterRemoved)
+					{
+						allKeyItems.Remove(Item.Floater);
+					}
+
+					var itemShopItem = SelectVendorItem(incentives.ToList(), nonincentives, treasurePool, incentiveLocationPool, rng);
+					placedItems.Add(new ItemShopSlot(_caravanItemLocation, itemShopItem));
+
+					allPlacements.Remove(itemShopItem);
+					allKeyItems.Remove(itemShopItem);
+
+					for (int cycle = 1; cycle <= 100; cycle++)
+					{
+						var (_, mapLocations, requirements) = _checker.CheckSanity(placedItems, null, _flags);
+						var accessibleSources = GetAllAccessibleRewardSources(preBlackOrbLocationPool, placedItems);
+
+						var incSourceCount = accessibleSources.Where(s => incentiveLocationPool.Contains(s)).Count();
+						var nonSourceCount = accessibleSources.Count - incSourceCount;
+
+						var testSource = accessibleSources.FirstOrDefault();
+						if (testSource == null)
+						{
+							placementFailed = true;
+							break;
+						}
+
+						List<(Item item, int incCount, int nonCount)> candidates = new List<(Item item, int incCount, int nonCount)>();
+
+						foreach (var item in allKeyItems)
+						{
+							placedItems.Add(NewItemPlacement(testSource, item));
+							_ = _checker.CheckSanity(placedItems, null, _flags);
+							placedItems.RemoveAt(placedItems.Count - 1);
```

Comment:
It's a sideeffect of the way the original SanityChecker was written. The SanityChecker gives 3 results. Complete, AccessRequirements and MapLocations. The first one isn't needed here(we know it's not complete). The other two are useless to the V2. Originally the MapLocations allow to look up lists of chests. But those Lists aren't fixed in V2, so the MapLocations can't be used. The AccessRequirements are used to check whether a Chest is accessible or not. But with V2 that's also not a static list anymore.
When I introduced the V2 I had to create an interface which changed as little as possible, so it would be clear the V2 doesn't have any sideeffects.

The V1 used a static method to make the chest lookup and this method took the return calues of the CheckSanity function. I changed it to be a member function and made the V1 into an own class. The V2 stores the acessible chests internally. CheckSanity is used tu update that list. IsRewardSourceAccessible is used to query it.

It's awkward and a thing that should be corrected now that V2 has been made the standard. But it's beyond the scope of this PR.